text/0000-fabric-gateway.md

```diff
@@ -0,0 +1,207 @@
+
+Publication of SDK releases will be as follows:
+- Node SDK published to NPM.
+- Java SDK published to Maven Central.
+- Go SDK published to separate GitHub repo (e.g. fabric-gateway-go) and tagged.
+
+Contributors are invited to create SDKs for other languages, e.g. Python, Rust.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+When run as a standalone process, the gateway would add another docker containter to an (arguably) already complex Fabric architecture.  This could be mitigated by embedding the gateway server within the organisation's existing peers.  This feature will be designed to be embeddable, if required.
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+- Why is this design the best in the space of possible designs?
+- What other designs have been considered and what is the rationale for not
+  choosing them?
+- What is the impact of not doing this?
+
+# Prior art
+[prior-art]: #prior-art
+
+The Gateway SDK has been available as the high level programming model in the Fabric SDKs since v1.4.  It has been thoroughly tested in manu scenarios.  This Fabric Gateway component is an embodiment of that programming model in a server component.
+
+# Testing
+[testing]: #testing
+
+In addition to the usual unit tests for all functions, a comprehensive end to end 
+scenario test suite will be created from the outset.  A set of 'cucumber' BDD style 
+tests will be taken from the existing Java and Node SDKs to test all the major functions
+of the gateway against a multi-org multi-peer network.  These tests will include:
+- evaluation and submission of transactions
+- event handling scenarios
+- discovery
+- transient / private data
+- error handling
+
+The CI pipeline will run the unit tests for the Gateway and its SDKs as well as the end-to-end scenario (cucumber) tests against all SDKs.
+
+# Dependencies
+[dependencies]: #dependencies
+
+- Fabric protobuf definitions will be pulled from the `hyperledger/fabric-protos` repo.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+### Client inspection of proposal responses
+
+Fabric is designed such that the client should have the opportunity to inspect the transaction response payload and then make a decision on whether or not that transaction should be sent to the orderer to be committed.
+
+There is the possibility of a client being able to send proposals that are subsequently not submitted to the orderer and recorded on the ledger, either because:
+- The proposal is not successfully endorsed; or
+- The client actively chooses not to submit a successfully endorsed transaction.
+
+This is viewed by some as a potential security vulnerability as it allows an attacker to probe the system.
+
+Fabric Gateway should provide the capability for some kind of audit of submitted proposals, including their transaction ID so they can be reconciled against committed transactions.
```

Comment:
How does a client authenticate to a gateway?

Reply:
I think that if you want client authentication, it's likely to be mutual TLS similar to the rest of Fabric. I guess things like OAuth are also a possibility. I don't think this is set in stone.

I think the statement you've commented on refers to the opportunity for the Gateway to log client interactions with the network rather than authentication.

Reply:
We've been talking about a new authentication service. We used mutual TLS for the new orderer channel management participation api as a temporary solution, with the expectation that we'd shift to a token-based authentication service as the strategic solution. I expect we would want to do the same for gateway. Look for a separate RFC in this space.